src/lib.rs

```diff
@@ -27,102 +18,783 @@ pub extern "C" fn wapc_init() {
     register_function("protocol_version", protocol_version_guest);
 }
 
+#[derive(Debug, PartialEq)]
+enum PolicyResponse {
+    Accept,
+    Reject(String),
+    Mutate(serde_json::Value),
+}
+
 fn validate(payload: &[u8]) -> CallResult {
     let validation_request: ValidationRequest<Settings> = ValidationRequest::new(payload)?;
 
-    info!(LOG_DRAIN, "starting validation");
-
-    // TODO: you can unmarshal any Kubernetes API type you are interested in
-    match serde_json::from_value::<apicore::Pod>(validation_request.request.object) {
-        Ok(pod) => {
-            // TODO: your logic goes here
-            if pod.metadata.name == Some("invalid-pod-name".to_string()) {
-                let pod_name = pod.metadata.name.unwrap();
-                info!(
-                    LOG_DRAIN,
-                    "rejecting pod";
-                    "pod_name" => &pod_name
-                );
-                kubewarden::reject_request(
-                    Some(format!("pod name {} is not accepted", &pod_name)),
-                    None,
-                )
+    let pod = match serde_json::from_value::<apicore::Pod>(validation_request.request.object) {
+        Ok(pod) => pod,
+        Err(_) => return kubewarden::accept_request(),
+    };
+
+    let settings = validation_request.settings;
+
+    match do_validate(pod, settings)? {
+        PolicyResponse::Accept => kubewarden::accept_request(),
+        PolicyResponse::Reject(message) => kubewarden::reject_request(Some(message), None),
+        PolicyResponse::Mutate(mutated_object) => kubewarden::mutate_request(mutated_object),
```

Comment:
Very nice approach. I would like to see this before implementing the `user-group-policy`. I could use something like this there.